pkg/capacity/topology/nodes.go

```diff
@@ -0,0 +1,328 @@
+			if unknown, ok := obj.(cache.DeletedFinalStateUnknown); ok && unknown.Obj != nil {
+				obj = unknown.Obj
+			}
+			csiNode, ok := obj.(*storagev1.CSINode)
+			if !ok {
+				klog.Errorf("deleted object: expected CSINode, got %T -> ignoring it", obj)
+				return
+			}
+			klog.V(5).Infof("capacity topology: removed CSINode: %s", csiNode.Name)
+			queue.Add("")
+		},
+	}
+	csiNodeInformer.Informer().AddEventHandler(csiNodeHandler)
+
+	return nt
+}
+
+var _ Informer = &nodeTopology{}
+
+type nodeTopology struct {
+	driverName      string
+	client          kubernetes.Interface
+	nodeInformer    coreinformersv1.NodeInformer
+	csiNodeInformer storageinformersv1.CSINodeInformer
+	queue           workqueue.RateLimitingInterface
+
+	mutex     sync.Mutex
+	segments  []*Segment
+	callbacks []Callback
+}
+
+// driverTopologyKeys returns nil if the driver is not running
+// on the node, otherwise at least an empty slice of topology keys.
+func (nt *nodeTopology) driverTopologyKeys(csiNode *storagev1.CSINode) []string {
+	for _, csiNodeDriver := range csiNode.Spec.Drivers {
+		if csiNodeDriver.Name == nt.driverName {
+			if csiNodeDriver.TopologyKeys == nil {
+				return []string{}
+			}
+			return csiNodeDriver.TopologyKeys
+		}
+	}
+	return nil
+}
+
+func (nt *nodeTopology) AddCallback(cb Callback) {
+	nt.mutex.Lock()
+	defer nt.mutex.Unlock()
+
+	nt.callbacks = append(nt.callbacks, cb)
+}
+
+func (nt *nodeTopology) List() []*Segment {
+	nt.mutex.Lock()
+	defer nt.mutex.Unlock()
+
+	// We need to return a new slice to protect against future
+	// changes in nt.segments. The segments themselves are
+	// immutable and shared.
+	segments := make([]*Segment, len(nt.segments))
+	copy(segments, nt.segments)
+	return segments
+}
+
+func (nt *nodeTopology) Run(ctx context.Context) {
+	go nt.nodeInformer.Informer().Run(ctx.Done())
+	go nt.csiNodeInformer.Informer().Run(ctx.Done())
+	go nt.runWorker(ctx)
+
+	klog.Info("Started node topology informer")
+	<-ctx.Done()
+	klog.Info("Shutting node topology informer")
+}
+
+func (nt *nodeTopology) HasSynced() bool {
+	if nt.nodeInformer.Informer().HasSynced() &&
+		nt.csiNodeInformer.Informer().HasSynced() {
+		// Now that both informers are up-to-date, use that
+		// information to update our own view of the world.
+		nt.sync(context.Background())
+		return true
+	}
+	return false
+}
+
+func (nt *nodeTopology) runWorker(ctx context.Context) {
+	for nt.processNextWorkItem(ctx) {
+	}
+}
+
+func (nt *nodeTopology) processNextWorkItem(ctx context.Context) bool {
+	obj, shutdown := nt.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer nt.queue.Done(obj)
+	nt.sync(ctx)
+	return true
+}
+
+func (nt *nodeTopology) sync(ctx context.Context) {
+	// For all nodes on which the driver is registered, collect the topology key/value pairs
+	// and sort them by key name to make the result deterministic. Skip all segments that have
+	// been seen before.
+	segments := nt.List()
+	removed := map[*Segment]bool{}
```

Comment:
Can you name this something like removalCandidates since everything in here may not be actually removed.